exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetGroupScan.java

```diff
@@ -219,229 +200,176 @@ public ScanStats getScanStats() {
-  public LogicalExpression getFilter() {
-    return filter;
-  }
-
-  public void setFilter(LogicalExpression filter) {
-    this.filter = filter;
+    return readEntries;
   }
 
   @Override
-  public AbstractParquetGroupScan applyFilter(LogicalExpression filterExpr, UdfUtilities udfUtilities,
+  public AbstractGroupScanWithMetadata applyFilter(LogicalExpression filterExpr, UdfUtilities udfUtilities,
       FunctionImplementationRegistry functionImplementationRegistry, OptionManager optionManager) {
-
-    if (!parquetTableMetadata.isRowGroupPrunable() ||
-        rowGroupInfos.size() > optionManager.getOption(PlannerSettings.PARQUET_ROWGROUP_FILTER_PUSHDOWN_PLANNING_THRESHOLD)) {
-      // Stop pruning for 2 cases:
-      //    -  metadata does not have proper format to support row group level filter pruning,
-      //    -  # of row groups is beyond PARQUET_ROWGROUP_FILTER_PUSHDOWN_PLANNING_THRESHOLD.
-      return null;
-    }
-
-    final Set<SchemaPath> schemaPathsInExpr = filterExpr.accept(new ParquetRGFilterEvaluator.FieldReferenceFinder(), null);
-
-    final List<RowGroupInfo> qualifiedRGs = new ArrayList<>(rowGroupInfos.size());
-
-    ParquetFilterPredicate filterPredicate = getParquetFilterPredicate(filterExpr, udfUtilities, functionImplementationRegistry, optionManager, true);
-
+    // Builds filter for pruning. If filter cannot be built, null should be returned.
+    FilterPredicate filterPredicate = getFilterPredicate(filterExpr, udfUtilities, functionImplementationRegistry, optionManager, true);
     if (filterPredicate == null) {
+      logger.debug("FilterPredicate cannot be built.");
       return null;
     }
 
-    boolean matchAllRowGroupsLocal = true;
-
-    for (RowGroupInfo rowGroup : rowGroupInfos) {
-      final ColumnExplorer columnExplorer = new ColumnExplorer(optionManager, columns);
-      List<String> partitionValues = getPartitionValues(rowGroup);
-      Map<String, String> implicitColValues = columnExplorer.populateImplicitColumns(rowGroup.getPath(), partitionValues, supportsFileImplicitColumns());
+    Set<SchemaPath> schemaPathsInExpr =
+        filterExpr.accept(new FilterEvaluatorUtils.FieldReferenceFinder(), null);
 
-      ParquetMetaStatCollector statCollector = new ParquetMetaStatCollector(
-          parquetTableMetadata,
-          rowGroup.getColumns(),
-          implicitColValues);
+    RowGroupScanFilterer builder = getFilterer().getFiltered(optionManager, filterPredicate, schemaPathsInExpr);
 
-      Map<SchemaPath, ColumnStatistics> columnStatisticsMap = statCollector.collectColStat(schemaPathsInExpr);
-
-      ParquetFilterPredicate.RowsMatch match = ParquetRGFilterEvaluator.matches(filterPredicate,
-          columnStatisticsMap, rowGroup.getRowCount(), parquetTableMetadata, rowGroup.getColumns(), schemaPathsInExpr);
-      if (match == ParquetFilterPredicate.RowsMatch.NONE) {
-        continue; // No row comply to the filter => drop the row group
+    if (getRowGroupsMetadata() != null) {
```

Comment:
Looks like _getRowGroupsMetadata()_ will never return a null. So are all the "else" clauses below dead code ?

Reply:
Good question! In the current code, it cannot be null since `ParquetMetadataProvider` cannot return null when calling `getRowGroupsMeta()` method, but in future will be provided other implementations of `TableMetadataProvider` which can return null for example for the case when the number of row groups is pretty large. So all the "else" will be suitable for such a case.

Reply:
Thanks @vvysotskyi ; at some point we could add your answer in a comment there.

Reply:
I have made some changes, so now for the case when metadata is not available, and it may be returned as a collection, an empty collection will be returned instead of null to reduce calls into `TableMetadataProvider`. Added the corresponding comment into `TableMetadataProvider` and replaced unneeded null checks by `!.isEmpty()` check in the last commit.